### PR TITLE
Place Lua script in CDATA section(s)

### PR DESCRIPTION
**Pull request by Legimet**
Note that Unicode escaping is no longer needed for Lua scripts since they go in a CDATA section.

This is true for CDATA sections in general, and the fact that escape_unicode does not skip CDATA sections is a bug.

### CODE REVIEW

File: luna.c [hunk 1]
```diff
@@ -121,34 +121,23 @@ void *escape_unicode(char *in_buf, size_t header_size, size_t footer_size, size_
...
-				memmove(p + 4, p + 1, in_buf + header_size + in_size + new_written - p - 1);
-				memcpy(p, "&lt;", 4);
-				new_written += 3;
+/* sub-routine of xml_compress() to fix occurrences of CDATA end sequence "]]>" in Lua scripts
+ * by spitting them between two CDATA sections. Returns the new in_buf, or NULL if out of memory. */
+void *fix_cdata_end_seq(char *in_buf, size_t header_size, size_t in_size, size_t *obuf_size) {
+	for (size_t offset = header_size; offset < header_size + in_size - 2; offset++) {
+		if (!memcmp(in_buf + offset, "]]>", 3)) {
+			*obuf_size += 12; // ]]><![CDATA[
+			char *tmp_in_buf;
```
Comment (Vogtinator) [inline luna.c:130]:
```
			char *new_in_buf;
```

File: luna.c [hunk 2]
```diff
@@ -121,34 +121,23 @@ void *escape_unicode(char *in_buf, size_t header_size, size_t footer_size, size_
...
-			} else if (*p == '<') {
-				memmove(p + 4, p + 1, in_buf + header_size + in_size + new_written - p - 1);
-				memcpy(p, "&lt;", 4);
-				new_written += 3;
+/* sub-routine of xml_compress() to fix occurrences of CDATA end sequence "]]>" in Lua scripts
+ * by spitting them between two CDATA sections. Returns the new in_buf, or NULL if out of memory. */
+void *fix_cdata_end_seq(char *in_buf, size_t header_size, size_t in_size, size_t *obuf_size) {
+	for (size_t offset = header_size; offset < header_size + in_size - 2; offset++) {
+		if (!memcmp(in_buf + offset, "]]>", 3)) {
+			*obuf_size += 12; // ]]><![CDATA[
```
Comment (Vogtinator) [inline luna.c:129]: Please have `"]]><![CDATA["` as string constant and use `sizeof - 1` or `strlen` instead of magic numbers

File: luna.c [hunk 3]
```diff
@@ -121,34 +121,23 @@ void *escape_unicode(char *in_buf, size_t header_size, size_t footer_size, size_
...
+			if (!(tmp_in_buf = realloc(in_buf, *obuf_size))) {
+				puts("can't realloc in_buf");
+				free(in_buf);
+				return NULL;
 			}
+			in_buf = tmp_in_buf;
+			memmove(in_buf + offset + 14, in_buf + offset + 2, header_size + in_size - (offset + 2));
+			memcpy(in_buf + offset + 2, "]]><![CDATA[", 12);
+			in_size += 12;
+			offset += 14;
```
Comment (Vogtinator) [inline luna.c:140]: This might have to be offset + 13, to account for the `offset++` every iteration. Could be verified by using `]]>]]>` as input.

**Comment by adriweb** [comment]
Any update @Legimet ?

File: luna.c (hunk 3 above)
Comment (Vogtinator) [reply luna.c]: Nope, the additional `offset++` actually skips past the `>`.

File: luna.c (hunk 2 above)
Comment (Vogtinator) [reply luna.c]: Applied

File: luna.c (hunk 1 above)
Comment (Vogtinator) [reply luna.c]: Applied

**Pull request merged**